### DISCUSSION TITLE
Set pythonpath in Github Actions, running on Ubuntu.

### DISCUSSION THREAD

**Discussion by dlesz** [community/community#25010] [Actions]
<p>I am relying on modules in my project for which I would like to set append a PYTHONPATH.</p>
<p>Working locally I would do it like this and it works.</p>
<pre><code>export PYTHONPATH=$PYTHONPATH:/home/&lt;user&gt;/&lt;code&gt;
</code></pre>
<p>However, the same approach for Github Actions throws the error:</p>
<pre><code>ModuleNotFoundError: No module named 'xxxx'
</code></pre>
<p>my_workflow.yml</p>
<pre><code>name: my_workflow

on:
push:
branches:
- master
pull_request:
branches:
- master
jobs:
build:

runs-on: ubuntu-latest

steps:
- uses: actions/checkout@v2
- name: Set up Python 3.8
uses: actions/setup-python@v1
with:
python-version: 3.8
- name: Install dependencies
run: |
python -m pip install --upgrade pip
pip install -r requirements.txt

- name: set pythonpath
run: |
# this does not work.
export PYTHONPATH=$PYTHONPATH:/home/runner/work/&lt;repo_name&gt;
- name: Test with pytest
run: |
pip install pytest
pytest
</code></pre>
<p>pwd returns</p>
<pre><code>/home/runner/work/&lt;repo_name&gt;/&lt;repo_name&gt;
</code></pre>
<p>I have tried to to set the PYTHONPATH to both </p>
<pre><code>.../work/&lt;repo_name&gt;
</code></pre>
<p>and the above, with no luck.</p>
<p>Please help!</p>

**Comment by chrispat** [comment]
<p>Each step in actions runs in its own process so exporting in a shell is not going to translate across steps.  To update the path for a downstream step you need to use the add-path workflow command <a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#adding-a-system-path" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#adding-a-system-path</a>.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>command: echo “::add-path::/path/to/dir” will pepend a directory to the system PATH variable for all subsequent actions in the current job. But look at your code, are you searching ‘set-env’?</p>
<pre><code>echo "::set-env name=PYTHONPATH::home/runner/work/&lt;repo_name&gt;"
</code></pre>

**Reply by arnodunstatter** [reply]
[GitHub Actions deprecated set-env and add-path commands](https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/) so instead the GITHUB_ENV variable should be used, as explained in the [documentation](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-system-path).

I'm having the same problem as dlesz, but what's strange is if I SSH into my runner (via the [tmate action](https://github.com/mxschmitt/action-tmate)) before I call the python code that has the ModuleNotFoundError I can run that file without error during the SSH session. 

Further, if I inspect the PYTHONPATH right before calling the file that produces the error, I can see the path was indeed added correctly. Really scratching my head on this one. 
```
      - name: updating PATH to enable importing my package
        run: echo "PYTHONPATH=/d/a/my_repo_name/my_repo_name/folder_containing_my_package" >> $GITHUB_ENV
        shell: bash

      - name: Run Tests
        run: |
          echo %PYTHONPATH%
          cd test_directory
          python test_which_imports_my_package.py
        shell: cmd
```


**UPDATE** My problem was I wasn't using the mounted path notation! Just needed to use `echo "PYTHONPATH=D:/a/my_repo_name/my_repo_name/folder_containing_my_package" >> $GITHUB_ENV` instead

**Comment by Mutepuka** [comment]
<p>Hello i keep having the same error. can you give me a step by set on how i solve this. </p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4c2b4591d77fe524014ce5d5875a66eb561e2d64.png" data-download-href="/uploads/short-url/aROZOIx8sWxTcok6XGc2PoiXCnO.png?dl=1" title="test_error.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/18482756/181088150-9bb132d4-a1cd-48a4-897a-e2a40f612dd0.png" alt="test_error.png" data-base62-sha1="aROZOIx8sWxTcok6XGc2PoiXCnO" width="690" height="387" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/4/4c2b4591d77fe524014ce5d5875a66eb561e2d64_2_690x387.png, https://aws1.discourse-cdn.com/github/optimized/2X/4/4c2b4591d77fe524014ce5d5875a66eb561e2d64_2_1035x580.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/4/4c2b4591d77fe524014ce5d5875a66eb561e2d64.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4c2b4591d77fe524014ce5d5875a66eb561e2d64_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">test_error.png</span><span class="informations">1366×768 156 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by lschmiddey** [comment]
<p>The set-env and add-path commands are deprecated in git, you can’t use them <a href="https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/" rel="noopener nofollow ugc">anymore</a>.<br>
Instead what you should do is set your environment variables like <a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#environment-files" rel="noopener nofollow ugc">this</a>:</p>
<blockquote>
<pre><code>- name: set pythonpath
  run: |
    echo "PYTHONPATH=mlprepare" &gt;&gt; $GITHUB_ENV
- name: Test with pytest
  run: |
    pytest</code></pre>
</blockquote>
